core/src/block.rs

```diff
@@ -420,7 +425,63 @@ mod valid {
             genesis_account: &AccountId,
             state_block: &mut StateBlock<'_>,
         ) -> WithEvents<Result<ValidBlock, (SignedBlock, BlockValidationError)>> {
-            let expected_block_height = state_block
+            if let Err(error) =
+                Self::validate_header(&block, topology, genesis_account, state_block)
+            {
+                return WithEvents::new(Err((block, error.into())));
+            }
+
+            if let Err(error) =
+                Self::validate_transactions(&block, expected_chain_id, genesis_account, state_block)
+            {
+                return WithEvents::new(Err((block, error.into())));
+            }
+
+            WithEvents::new(Ok(ValidBlock(block)))
+        }
+
+        /// Same as `validate` but:
+        /// * Block header will be validated with read-only state
+        /// * If block header is valid, `voting_block` will be released,
+        ///   and transactions will be validated with write state
+        pub fn validate_keep_voting_block<'state>(
+            block: SignedBlock,
+            topology: &Topology,
+            expected_chain_id: &ChainId,
+            genesis_account: &AccountId,
+            state: &'state State,
+            voting_block: &mut Option<VotingBlock>,
+        ) -> WithEvents<Result<(ValidBlock, StateBlock<'state>), (SignedBlock, BlockValidationError)>>
+        {
+            if let Err(error) =
+                Self::validate_header(&block, topology, genesis_account, &state.view())
+            {
+                return WithEvents::new(Err((block, error)));
+            }
+
+            // Release block writer before creating new one
+            let _ = voting_block.take();
+            let mut state_block = state.block();
+
+            if let Err(error) = Self::validate_transactions(
```

Comment:
Cases when voting block going to be discarded falsely:

Case 1:
1. peer already have voting block
2. new block is received through block created message
3. this block has proper height, matching topology, proper leader signature
4. BUT incorrect transactions in the new block

This means that leader send two blocks.

I don't think this is issue because after such event current leader would loose it's place in the validators set. 

And no one other than current leader can perform this operation.

Case 2:
1. peer already have voting block
2. new block received through block sync message
3. this block has proper height (for soft-fork or ordinary condition), view change index larger than current one (!!!), proper leader signature for this larger signature
4. here is two options why this block might be incorrect:
    - incorrect transactions
    - missing proxy tail signature, not enough voting peers signatures

In case when peer conclude that transactions are incorrect but block has all valid signatures indicates that something is certainly wrong with peer itself.

Case when block has missing signatures (and has incorrect transactions) opens possibility for the following attack by any peer in the topology:
1. find view change index (higher than current once)
2. create block 
3. sign it
4. send it through block sync to other peers
5. peers will discard current voting block and this block going to be rejected
6. current view change index will fail 
7. repeat the same process for the new view change index

To mitigate this issue we could check all signatures before trying to execute transactions for the given block (and discarding previous voting block).

I think this can be done in separate ticket or in the current one as you see fit.

Reply:
#4834